draft.md

```diff
@@ -0,0 +1,99 @@
+# Draft IPLD Unixfs Spec
+
+## Basic Structure
+
+A Unixfs is either a file or a directory.
+The top level IPLD object is a CBOR map with at least two fields: `type` and `data`
+and maybe a few other such as a version string or a set of flags.
+The `type` field is either `file` or `dir`.
+
+## IPLD `file`
+
+If an IPLD file is a leaf its CID type is `raw` (0x55) and has no structure.
+Otherwise its CID type is `dag-cbor` (0x71).
+The `type` field is set to `file` and the `data` field is an CBOR array.
+Each element of the array is CBOR map with the following fields:
+
+  - `data`: link
+  - `size`: cumulative size of `data`
+  - `fsize`: (file size) cumulative size of the payload of `data`
+  
+The `fsize` field is omitted if the link is `raw` as it is the same value as size.
+
+## IPLD `dir`
+
+An IPLD `dir` represents a directory.
+Its CID type is `dag-cbor` (0x71).
+The `type` field set to `dir` and the data field is an CBOR map.
+The key of the map is a filename and is a CBOR text string encoded in UTF-8.
+The value of the map is another CBOR map with the following standard fields:
+
+  - `type`
+  - `exe`: CBOR boolean: executable bit
+  - `data`: normally a CBOR link, but can be other types depending on the value of the `type` field
+  - `size`: cumulative size of `data`
+  - `fsize`: (file size) cumulative size of the payload of `data`
+  - `fname`: CBOR byte string: original filename if it differs from the key
+
+And at least the following optional fields:
+
+  - `ro`: CBOR boolean: read only
+  - `mtime`: Modification time
+  - `attr`: CBOR Map: Extended attributes
+
+Additional fields may be defined.  All implementation specific or user
+defined fields should be stored under the `attr` field. 
+
+### Directory Types
+
+The type field is limited to a set of well defined values:
+
+  * _omitted_: regular file 
+  * `dir`: directory entry
+  * `special`: special file type (fifo, device, etc).
+     The `data` field is a CBOR Map with at least one field to describe the type.
```

Comment:
I don't think we should overload 'data', especially avoid making it have different types based on the value of a key in the parent level. That sort of parsing is *hard* to do efficiently

Reply:
@whyrusleeping 

I like the similarity of having the contents of the file entry always be in the same field, for most types it is an IPLD link, for symbolic links it is the target, for special file types in a CBOR map with the details of the special file.  My thinking was the type would just be an interface and then cast the correct type once it is known.

I can instead have the following fields:

* `link`: CBOR link when applicable
* `target`: symbolic link target
* `data`: a CBOR map that contains additional data for the directory entry that is not a `link` or `target`

I rather not provide special fields to describe all the different types of `special` files.

Reply:
we're going to have to inspect fields to determine what to do with things anyways. Overloading things doesnt really save us much in my opinion.

Reply:
@whyrusleeping I am having a hard time interpreting that comment, are you okay with my proposal (`link`, `target`, `data` fields) are you saying we should create special fields for each and every special file type?

Reply:
More to the point, I don't want to enumerate the required fields in the first version of the spec.  I like the `data` field abstraction because we can defer that part for later, it also allows us to to define a standard set of tags so that an implementation can error out of it find something it doesn't understand.